// @aws-cdk/assets-docker/lib/image-asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const cxapi = require("@aws-cdk/cx-api");
const fs = require("fs");
const path = require("path");
const adopted_repository_1 = require("./adopted-repository");
/**
 * An asset that represents a Docker image.
 *
 * The image will be created in build time and uploaded to an ECR repository.
 */
class DockerImageAsset extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        // resolve full path
        this.directory = path.resolve(props.directory);
        if (!fs.existsSync(this.directory)) {
            throw new Error(`Cannot find image directory at ${this.directory}`);
        }
        if (!fs.existsSync(path.join(this.directory, 'Dockerfile'))) {
            throw new Error(`No 'Dockerfile' found in ${this.directory}`);
        }
        const imageNameParameter = new cdk.CfnParameter(this, 'ImageName', {
            type: 'String',
            description: `ECR repository name and tag asset "${this.node.path}"`,
        });
        const asset = {
            packaging: 'container-image',
            path: this.directory,
            id: this.node.uniqueId,
            imageNameParameter: imageNameParameter.logicalId,
            repositoryName: props.repositoryName,
        };
        this.node.addMetadata(cxapi.ASSET_METADATA, asset);
        // parse repository name and tag from the parameter (<REPO_NAME>:<TAG>)
        const components = cdk.Fn.split(':', imageNameParameter.stringValue);
        const repositoryName = cdk.Fn.select(0, components).toString();
        const imageTag = cdk.Fn.select(1, components).toString();
        // Require that repository adoption happens first, so we route the
        // input ARN into the Custom Resource and then get the URI which we use to
        // refer to the image FROM the Custom Resource.
        //
        // If adoption fails (because the repository might be twice-adopted), we
        // haven't already started using the image.
        this.repository = new adopted_repository_1.AdoptedRepository(this, 'AdoptRepository', { repositoryName });
        this.imageUri = this.repository.repositoryUriForTag(imageTag);
    }
}
exports.DockerImageAsset = DockerImageAsset;
