// @aws-cdk/aws-servicediscovery/lib/cname-instance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const instance_1 = require("./instance");
const namespace_1 = require("./namespace");
const service_1 = require("./service");
const servicediscovery_generated_1 = require("./servicediscovery.generated");
/*
 * Instance that is accessible using a domain name (CNAME).
 */
class CnameInstance extends instance_1.InstanceBase {
    constructor(scope, id, props) {
        super(scope, id);
        if (props.service.namespace.type === namespace_1.NamespaceType.Http) {
            throw new Error('Namespace associated with Service must be a DNS Namespace.');
        }
        if (props.service.dnsRecordType !== service_1.DnsRecordType.CNAME) {
            throw new Error('A `CnameIntance` can only be used with a service using a `CNAME` record.');
        }
        const resource = new servicediscovery_generated_1.CfnInstance(this, 'Resource', {
            instanceId: props.instanceId || this.uniqueInstanceId(),
            serviceId: props.service.serviceId,
            instanceAttributes: Object.assign({ AWS_INSTANCE_CNAME: props.instanceCname }, props.customAttributes)
        });
        this.service = props.service;
        this.instanceId = resource.instanceId;
        this.instanceCname = props.instanceCname;
    }
}
exports.CnameInstance = CnameInstance;
