// @aws-cdk/aws-servicediscovery/lib/namespace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
var NamespaceType;
(function (NamespaceType) {
    /**
     * Choose this option if you want your application to use only API calls to discover registered instances.
     */
    NamespaceType["Http"] = "HTTP";
    /**
     * Choose this option if you want your application to be able to discover instances using either API calls or using
     * DNS queries in a VPC.
     */
    NamespaceType["DnsPrivate"] = "DNS_PRIVATE";
    /**
     * Choose this option if you want your application to be able to discover instances using either API calls or using
     * public DNS queries. You aren't required to use both methods.
     */
    NamespaceType["DnsPublic"] = "DNS_PUBLIC";
})(NamespaceType = exports.NamespaceType || (exports.NamespaceType = {}));
class NamespaceBase extends cdk.Construct {
    export() {
        return {
            namespaceName: new cdk.CfnOutput(this, 'NamespaceName', { value: this.namespaceArn }).makeImportValue().toString(),
            namespaceArn: new cdk.CfnOutput(this, 'NamespaceArn', { value: this.namespaceArn }).makeImportValue().toString(),
            namespaceId: new cdk.CfnOutput(this, 'NamespaceId', { value: this.namespaceId }).makeImportValue().toString(),
            type: this.type,
        };
    }
}
exports.NamespaceBase = NamespaceBase;
// The class below exists purely so that users can still type Namespace.import().
// It does not make sense to have HttpNamespace.import({ ..., type: NamespaceType.PublicDns }),
// but at the same time ecs.Cluster wants a type-generic export()/import(). Hence, we put
// it in Namespace.
/**
 * Static Namespace class
 */
class Namespace {
    /**
     * Import a namespace
     */
    static import(scope, id, props) {
        return new ImportedNamespace(scope, id, props);
    }
    constructor() {
    }
}
exports.Namespace = Namespace;
class ImportedNamespace extends NamespaceBase {
    constructor(scope, id, props) {
        super(scope, id);
        this.namespaceId = props.namespaceId;
        this.namespaceArn = props.namespaceArn;
        this.namespaceName = props.namespaceName;
        this.type = props.type;
    }
}
