// @aws-cdk/aws-servicediscovery/lib/private-dns-namespace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const namespace_1 = require("./namespace");
const service_1 = require("./service");
const servicediscovery_generated_1 = require("./servicediscovery.generated");
/**
 * Define a Service Discovery HTTP Namespace
 */
class PrivateDnsNamespace extends namespace_1.NamespaceBase {
    constructor(scope, id, props) {
        super(scope, id);
        if (props.vpc === undefined) {
            throw new Error(`VPC must be specified for PrivateDNSNamespaces`);
        }
        const ns = new servicediscovery_generated_1.CfnPrivateDnsNamespace(this, 'Resource', {
            name: props.name,
            description: props.description,
            vpc: props.vpc.vpcId
        });
        this.namespaceName = props.name;
        this.namespaceId = ns.privateDnsNamespaceId;
        this.namespaceArn = ns.privateDnsNamespaceArn;
        this.type = namespace_1.NamespaceType.DnsPrivate;
    }
    /**
     * Creates a service within the namespace
     */
    createService(id, props) {
        return new service_1.Service(this, id, Object.assign({ namespace: this }, props));
    }
}
exports.PrivateDnsNamespace = PrivateDnsNamespace;
