// @aws-cdk/aws-servicediscovery/lib/public-dns-namespace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const namespace_1 = require("./namespace");
const service_1 = require("./service");
const servicediscovery_generated_1 = require("./servicediscovery.generated");
/**
 * Define a Public DNS Namespace
 */
class PublicDnsNamespace extends namespace_1.NamespaceBase {
    constructor(scope, id, props) {
        super(scope, id);
        const ns = new servicediscovery_generated_1.CfnPublicDnsNamespace(this, 'Resource', {
            name: props.name,
            description: props.description,
        });
        this.namespaceName = props.name;
        this.namespaceId = ns.publicDnsNamespaceId;
        this.namespaceArn = ns.publicDnsNamespaceArn;
        this.type = namespace_1.NamespaceType.DnsPublic;
    }
    /**
     * Creates a service within the namespace
     */
    createService(id, props) {
        return new service_1.Service(this, id, Object.assign({ namespace: this }, props));
    }
}
exports.PublicDnsNamespace = PublicDnsNamespace;
