// @aws-cdk/aws-ecs/lib/load-balanced-ecs-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ec2_service_1 = require("./ec2/ec2-service");
const ec2_task_definition_1 = require("./ec2/ec2-task-definition");
const load_balanced_service_base_1 = require("./load-balanced-service-base");
/**
 * A single task running on an ECS cluster fronted by a load balancer
 */
class LoadBalancedEc2Service extends load_balanced_service_base_1.LoadBalancedServiceBase {
    constructor(scope, id, props) {
        super(scope, id, props);
        const taskDefinition = new ec2_task_definition_1.Ec2TaskDefinition(this, 'TaskDef', {});
        const container = taskDefinition.addContainer('web', {
            image: props.image,
            memoryLimitMiB: props.memoryLimitMiB,
            environment: props.environment,
        });
        container.addPortMappings({
            containerPort: props.containerPort || 80,
        });
        const service = new ec2_service_1.Ec2Service(this, "Service", {
            cluster: props.cluster,
            desiredCount: props.desiredCount || 1,
            taskDefinition,
        });
        this.addServiceAsTarget(service);
    }
}
exports.LoadBalancedEc2Service = LoadBalancedEc2Service;
