// @aws-cdk/aws-ecs/lib/load-balanced-fargate-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const aws_route53_1 = require("@aws-cdk/aws-route53");
const fargate_service_1 = require("./fargate/fargate-service");
const fargate_task_definition_1 = require("./fargate/fargate-task-definition");
const load_balanced_service_base_1 = require("./load-balanced-service-base");
const aws_log_driver_1 = require("./log-drivers/aws-log-driver");
/**
 * A Fargate service running on an ECS cluster fronted by a load balancer
 */
class LoadBalancedFargateService extends load_balanced_service_base_1.LoadBalancedServiceBase {
    constructor(scope, id, props) {
        super(scope, id, props);
        const taskDefinition = new fargate_task_definition_1.FargateTaskDefinition(this, 'TaskDef', {
            memoryMiB: props.memoryMiB,
            cpu: props.cpu
        });
        const optIn = props.createLogs !== undefined ? props.createLogs : true;
        const container = taskDefinition.addContainer('web', {
            image: props.image,
            logging: optIn ? this.createAWSLogDriver(this.node.id) : undefined,
            environment: props.environment
        });
        container.addPortMappings({
            containerPort: props.containerPort || 80,
        });
        const assignPublicIp = props.publicTasks !== undefined ? props.publicTasks : false;
        const service = new fargate_service_1.FargateService(this, "Service", {
            cluster: props.cluster,
            desiredCount: props.desiredCount || 1,
            taskDefinition,
            assignPublicIp
        });
        this.service = service;
        this.addServiceAsTarget(service);
        if (typeof props.domainName !== 'undefined') {
            if (typeof props.domainZone === 'undefined') {
                throw new Error('A Route53 hosted domain zone name is required to configure the specified domain name');
            }
            new aws_route53_1.AliasRecord(this, "DNS", {
                zone: props.domainZone,
                recordName: props.domainName,
                target: this.loadBalancer
            });
        }
    }
    createAWSLogDriver(prefix) {
        return new aws_log_driver_1.AwsLogDriver(this, 'Logging', { streamPrefix: prefix });
    }
}
exports.LoadBalancedFargateService = LoadBalancedFargateService;
