// @aws-cdk/aws-ecs/lib/ec2/ec2-event-rule-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/cdk");
const util_1 = require("../util");
/**
 * Start a service on an EC2 cluster
 */
class Ec2EventRuleTarget extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        if (!util_1.isEc2Compatible(props.taskDefinition.compatibility)) {
            throw new Error('Supplied TaskDefinition is not configured for compatibility with EC2');
        }
        this.cluster = props.cluster;
        this.taskDefinition = props.taskDefinition;
        this.taskCount = props.taskCount !== undefined ? props.taskCount : 1;
    }
    /**
     * Allows using containers as target of CloudWatch events
     */
    asEventRuleTarget(_ruleArn, _ruleUniqueId) {
        const role = this.eventsRole;
        role.addToPolicy(new iam.PolicyStatement()
            .addAction('ecs:RunTask')
            .addResource(this.taskDefinition.taskDefinitionArn)
            .addCondition('ArnEquals', { "ecs:cluster": this.cluster.clusterArn }));
        return {
            id: this.node.id,
            arn: this.cluster.clusterArn,
            roleArn: role.roleArn,
            ecsParameters: {
                taskCount: this.taskCount,
                taskDefinitionArn: this.taskDefinition.taskDefinitionArn
            }
        };
    }
    /**
     * Create or get the IAM Role used to start this Task Definition.
     *
     * We create it under the TaskDefinition object so that if we have multiple EventTargets
     * they can reuse the same role.
     */
    get eventsRole() {
        let role = this.taskDefinition.node.tryFindChild('EventsRole');
        if (role === undefined) {
            role = new iam.Role(this.taskDefinition, 'EventsRole', {
                assumedBy: new iam.ServicePrincipal('events.amazonaws.com')
            });
        }
        return role;
    }
    /**
     * Prepare the Event Rule Target
     */
    prepare() {
        // If it so happens that a Task Execution Role was created for the TaskDefinition,
        // then the CloudWatch Events Role must have permissions to pass it (otherwise it doesn't).
        //
        // It never needs permissions to the Task Role.
        if (this.taskDefinition.executionRole !== undefined) {
            this.taskDefinition.executionRole.grantPassRole(this.eventsRole);
        }
    }
}
exports.Ec2EventRuleTarget = Ec2EventRuleTarget;
