// @aws-cdk/aws-secretsmanager/test/integ.secret.lit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iam = require("@aws-cdk/aws-iam");
const cdk = require("@aws-cdk/cdk");
const secretsManager = require("../lib");
class SecretsManagerStack extends cdk.Stack {
    constructor(scope, id) {
        super(scope, id);
        const role = new iam.Role(this, 'TestRole', { assumedBy: new iam.AccountRootPrincipal() });
        /// !show
        // Default secret
        const secret = new secretsManager.Secret(this, 'Secret');
        secret.grantRead(role);
        new iam.User(this, 'User', {
            password: secret.secretValue
        });
        // Templated secret
        const templatedSecret = new secretsManager.Secret(this, 'TemplatedSecret', {
            generateSecretString: {
                secretStringTemplate: JSON.stringify({ username: 'user' }),
                generateStringKey: 'password'
            }
        });
        new iam.User(this, 'OtherUser', {
            userName: templatedSecret.secretJsonValue('username').toString(),
            password: templatedSecret.secretJsonValue('password')
        });
        /// !hide
    }
}
const app = new cdk.App();
new SecretsManagerStack(app, 'Integ-SecretsManager-Secret');
app.run();
