// @aws-cdk/aws-ecs/test/ec2/test.ec2-event-rule-target.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const ec2 = require("@aws-cdk/aws-ec2");
const events = require("@aws-cdk/aws-events");
const cdk = require("@aws-cdk/cdk");
const ecs = require("../../lib");
module.exports = {
    "Can use EC2 taskdef as EventRule target"(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const vpc = new ec2.VpcNetwork(stack, 'Vpc', { maxAZs: 1 });
        const cluster = new ecs.Cluster(stack, 'EcsCluster', { vpc });
        cluster.addCapacity('DefaultAutoScalingGroup', {
            instanceType: new ec2.InstanceType('t2.micro')
        });
        const taskDefinition = new ecs.Ec2TaskDefinition(stack, 'TaskDef');
        taskDefinition.addContainer('TheContainer', {
            image: ecs.ContainerImage.fromRegistry('henk'),
            memoryLimitMiB: 256
        });
        const rule = new events.EventRule(stack, 'Rule', {
            scheduleExpression: 'rate(1 minute)',
        });
        // WHEN
        const target = new ecs.Ec2EventRuleTarget(stack, 'EventTarget', {
            cluster,
            taskDefinition,
            taskCount: 1
        });
        rule.addTarget(target, {
            jsonTemplate: {
                argument: 'hello'
            }
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Events::Rule', {
            Targets: [
                {
                    Arn: { "Fn::GetAtt": ["EcsCluster97242B84", "Arn"] },
                    EcsParameters: {
                        TaskCount: 1,
                        TaskDefinitionArn: { Ref: "TaskDef54694570" }
                    },
                    Id: "EventTarget",
                    InputTransformer: {
                        InputTemplate: "{\"argument\":\"hello\"}"
                    },
                    RoleArn: { "Fn::GetAtt": ["TaskDefEventsRoleFB3B67B8", "Arn"] }
                }
            ]
        }));
        test.done();
    }
};
